packages/calypso-e2e/src/jest-playwright-config/environment.ts

```diff
@@ -0,0 +1,198 @@
+/* eslint-disable require-jsdoc */
+import fs from 'fs/promises';
+import os from 'os';
+import path from 'path';
+import { Context } from 'vm';
+import JestEnvironmentNode from 'jest-environment-node';
+import { Browser, BrowserContext, BrowserContextOptions, chromium, Page } from 'playwright';
+import env from '../env-variables';
+import config from './playwright-config';
+import type { Config, Circus } from 'jest-environment-node/node_modules/@jest/types';
+
+const sanitizeString = ( text: string ) => {
+	return text.replace( /[^a-z0-9]/gi, '-' ).toLowerCase();
+};
+
+class JestEnvironmentPlaywright extends JestEnvironmentNode {
+	private testFilename: string;
+	private testArtifactsPath: string;
+	private failure?: {
+		type: 'hook' | 'test';
+		name: string;
+	};
+
+	constructor( config: Config.ProjectConfig, context: Context ) {
+		super( config );
+
+		this.testFilename = path.parse( context.testPath ).name;
+		this.testArtifactsPath = '';
+	}
+
+	/**
+	 * Set up the environment.
+	 */
+	async setup() {
+		await super.setup();
+
+		// Create folders for test artefacts.
+		await fs.mkdir( env.ARTIFACTS_PATH, { recursive: true } );
+		this.testArtifactsPath = await fs.mkdtemp(
+			path.join( env.ARTIFACTS_PATH, `${ this.testFilename }__${ Date.now() }-` )
+		);
+
+		const logFilePath = path.join( this.testArtifactsPath, `${ this.testFilename }.log` );
+
+		// Start the browser.
+		const browser = await chromium.launch( {
+			...config.launchOptions,
+			logger: {
+				log: async ( name: string, severity: string, message: string ) => {
+					await fs.appendFile(
+						logFilePath,
+						`${ new Date().toISOString() } ${ process.pid } ${ name } ${ severity }: ${ message }\n`
+					);
+				},
+				isEnabled: ( name ) => name === 'api',
+			},
+		} );
+
+		// Proxy our E2E browser to bind our custom default context options.
+		// We need to it this way because we're not using the first-party Playwright
+		// and we can't use the playwright.config.js file.
```

Comment:
Hm, interesting! Why aren't we using the first-party Playwright lib and what are we using instead?

Reply:
Ah, `jest-playwright`, I assume.

Reply:
Sergio and I had this discussion when he was still with a8c and IIRC the gist was that:
- Jest supports custom reporters while Playwright Test Runner does not;
- We can use other plugins with Jest (eg. Jest Runner Groups) whereas Playwright Test Runner won't.

I do have an item in the backlog, for a 20% time project, to play with the Playwright Test Runner. Its advantages are:
- first-party support for managing multiple pages/contexts and artifacts (screenshot, video) for those.
- interfaces well with simulated devices that Playwright ships with.
- much easier to set up cross-browser testing.

Reply:
Thanks a lot for elaborating :)